portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```
@@ -11,22 +11,80 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query(COMMENTS_ENTITY_NAME).addSort(TIMESTAMP_PROPERTY_KEY, SortDirection.DESCENDING);
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    
+    List<Post> comments = new ArrayList<>();
+    FetchOptions fetchOptions = FetchOptions.Builder.withLimit(max);
+    for (Entity entity : results.asIterable(fetchOptions)) {
+      long id = entity.getKey().getId();
+      String comment = (String) entity.getProperty(COMMENT_PROPERTY_KEY);
+      long timestamp = (long) entity.getProperty(TIMESTAMP_PROPERTY_KEY);
+     
+      Post post = new Post(id, comment, timestamp);
+      comments.add(post);
+     
+    }
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.getWriter().println(gson.toJson(comments));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userComment = request.getParameter(COMMENT_PARAMETER_NAME);
+    long timestamp = System.currentTimeMillis();
+    String maxString = request.getParameter(MAX_COMMENTS_PARAMETER_NAME);
+    if (userComment != null) {
+      Entity taskEntity = new Entity(COMMENTS_ENTITY_NAME);
+      taskEntity.setProperty(COMMENT_PROPERTY_KEY, userComment);
+      taskEntity.setProperty(TIMESTAMP_PROPERTY_KEY, timestamp);
+
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      datastore.put(taskEntity);
+    } else if (maxString != null) {
+      try {
+        max = Integer.parseInt(maxString);
+
+      } catch (NumberFormatException e) {
+        System.out.println("Could not convert to int: " + maxString);
```

Note: this generates a local system (AppEngine) log and swallows the exception. Do you want to surface this error to the user in any way?